src/vm_cpu.rs

```diff
@@ -96,6 +99,223 @@ pub struct VmCpuTrapState {
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<()> {
+        use PmuCounterState::*;
+        let new_state = CounterState {
+            config_flags,
+            event_type,
+            event_data,
+            value: 0,
+        };
+        let pmu_info = pmu::PmuInfo::get()?;
+        // Bail out if platform configuration returned a spurious counter index.
+        let _ = pmu_info
+            .get_counter_info(platform_counter_index)
+            .map_err(|_| SbiError::NotSupported)?;
+        let state = &mut self.counter_state[platform_counter_index as usize];
+        match state {
+            // If skip_match is set, counter must already be configured
+            Configured(c) | Started(c) if config_flags.is_skip_match() => {
+                if config_flags.is_auto_start() {
+                    *state = Started(*c);
+                }
+                Ok(())
+            }
+            NotConfigured if !config_flags.is_skip_match() => {
+                *state = if config_flags.is_auto_start() {
+                    Started(new_state)
+                } else {
+                    Configured(new_state)
+                };
+                Ok(())
+            }
+            _ => Err(SbiError::InvalidParam),
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to start counters.
+    pub fn update_started_counters(&mut self, counter_index: u64, counter_mask: u64) {
+        use PmuCounterState::*;
+        let bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+        for i in bitmask_iter {
+            if let Configured(c) = self.counter_state[i] {
+                self.counter_state[i] = Started(c);
+            }
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to stop counters.
+    pub fn update_stopped_counters(
+        &mut self,
+        counter_index: u64,
+        counter_mask: u64,
+        stop_flags: PmuCounterStopFlags,
+    ) {
+        use PmuCounterState::*;
+        let bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+        for i in bitmask_iter {
+            let state = &mut self.counter_state[i];
+            match state {
+                // Deliberately more permissive since the implementation permits
+                // operations even on stopped counters (example: stop_flag_reset).
+                Configured(c) | Started(c) => {
+                    if stop_flags.is_reset_flag() {
+                        *state = NotConfigured;
+                    } else {
+                        *state = Configured(*c);
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Returns a filtered counter_mask if the PMU counter range can be started.
+    pub fn is_counter_range_startable(
```

Comment:
Since these functions are now returning the filtered mask rather than performing a simple boolean check, it probably makes more sense to call them something like `startable_counter_range()` or `get_startable_counter_range()`, etc.